common/src/main/java/io/netty/util/AbstractReferenceCounted.java

```diff
@@ -77,16 +112,36 @@ public boolean release(int decrement) {
     }
 
     private boolean release0(int decrement) {
-        int oldRef = refCntUpdater.getAndAdd(this, -decrement);
-        if (oldRef == decrement) {
-            deallocate();
-            return true;
-        } else if (oldRef < decrement || oldRef - decrement > oldRef) {
-            // Ensure we don't over-release, and avoid underflow.
-            refCntUpdater.getAndAdd(this, decrement);
-            throw new IllegalReferenceCountException(oldRef, -decrement);
+        int rawCnt = nonVolatileRawCnt();
+        for (boolean firstTry = true;; firstTry = false) {
+            if ((rawCnt & 1) != 0) {
+                throw new IllegalReferenceCountException(0, -decrement);
+            }
+            int realCnt = rawCnt >>> 1;
+            if (decrement == realCnt) {
+                // most likely case
+                if (refCntUpdater.compareAndSet(this, rawCnt, 1)) { // any odd number
+                    deallocate();
+                    return true;
+                }
+                if (!firstTry) {
+                    // this benefits throughput under high contention
+                    Thread.yield();
+                }
+            } else if (decrement < realCnt) {
+                // all changes to the raw count are 2x the "real" change
+                if (refCntUpdater.compareAndSet(this, rawCnt, rawCnt - (decrement << 1))) {
+                    return false;
+                }
+                if (!firstTry) {
+                    // this benefits throughput under high contention
+                    Thread.yield();
+                }
+            } else if (!firstTry) {
+                throw new IllegalReferenceCountException(realCnt, -decrement);
+            }
+            rawCnt = refCntUpdater.get(this);
         }
```

Comment:
nit: consider simplify this:

```java
        for (boolean firstTry = true, int rawCnt = nonVolatileRawCnt();; firstTry = false, rawCnt = refCntUpdater.get(this)) {
             if ((rawCnt & 1) != 0) {
                 throw new IllegalReferenceCountException(0, -decrement);
             }
             int realCnt = rawCnt >>> 1;
             if (decrement == realCnt) {
                 // most likely case
                 if (refCntUpdater.compareAndSet(this, rawCnt, 1)) { // any odd number
                     deallocate();
                     return true;
                 }
             } else if (decrement < realCnt) {
                 // all changes to the raw count are 2x the "real" change
                 if (refCntUpdater.compareAndSet(this, rawCnt, rawCnt - (decrement << 1))) {
                     return false;
                 }
             } else if (!firstTry) {
                 throw new IllegalReferenceCountException(realCnt, -decrement);
             }
             if (!firstTry) {
                 // this benefits throughput under high contention
                 Thread.yield();
             }
             rawCnt = refCntUpdater.get(this);
         }
```

Reply:
This should also make the method size a bit smaller and so help with inlining.

Reply:
Same is true for `AbstractByteBuf`

Reply:
@normanmaurer lol I actually had the `rawCnt = refCntUpdater.get(this)` inside the `for` statement like that originally but thought better of it readability-wise (I have a bad habit of making things too compact), happy to do it though (I'm assuming you also meant to remove it inside the loop).

I don't think `refCnt` initialization can go in there though since the types have to match if there are multiple.

I didn't move out the `Thread.yield()` like that since it results in an additional check on one of the paths: when `firstTry` is true and `decrement > realCount`, which could happen in non-error cases especially since the first read is nonvolatile. But I guess probably uncommon enough to justify the simplification, I will do it.

I was also still planning to separate the less common paths although this will likely require some more duplication.

Reply:
@njhill I see... so yeah maybe just do the `firstTry` change + the split into methods. Thanks a lot  for all your effort!